gk/fib.c

```diff
@@ -1893,6 +1893,248 @@ l_list_gk_fib6(lua_State *l)
 	return 1;
 }
 
+#define CTYPE_STRUCT_ETHER_CACHE_PTR "struct ether_cache *"
+
+static void
+list_hash_table_neighbors(lua_State *l, struct neighbor_hash_table *neigh_ht)
+{
+	uint32_t next = 0;
+	const void *key;
+	void *data;
+	int32_t index;
+
+	static bool assigned_type_ether_cache = false;
+	static uint32_t correct_ctypeid_ether_cache;
+
+	void *cdata;
+
+	if (!assigned_type_ether_cache) {
+		correct_ctypeid_ether_cache = luaL_get_ctypeid(l,
+			CTYPE_STRUCT_ETHER_CACHE_PTR);
+		assigned_type_ether_cache = true;
+	}
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		lua_pushvalue(l, 2);
+		lua_insert(l, 3);
+		cdata = luaL_pushcdata(l, correct_ctypeid_ether_cache,
+			sizeof(struct ether_cache *));
+		*(struct ether_cache **)cdata = &neigh_ht->cache_tbl[index];
+		lua_insert(l, 4);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
+			/*
+			 * The following call isn't strictly necessary
+			 * because we are going to raise a Lua error.
+			 */
+			lua_error(l);
+		}
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+}
+
+static void
+list_ipv4_if_neighbors(lua_State *l,
+	struct gatekeeper_if *iface, struct gk_lpm *ltbl)
+{
+	char ip[128];
+	char fib_info[512];
+	struct gk_fib *neigh_fib;
+	int ret;
+	int fib_id = lpm_lookup_ipv4(ltbl->lpm, iface->ip4_addr.s_addr);
+	/*
+	 * Invalid gateway entry, since at least we should
+	 * obtain the FIB entry for the neighbor table.
+	 */
+	if (fib_id < 0)
+		luaL_error(l, "gk: failed to lookup the lpm table at %s!",
+			__func__);
+
+	neigh_fib = &ltbl->fib_tbl[fib_id];
+
+	if (inet_ntop(AF_INET, &iface->ip4_addr, ip, sizeof(ip)) == NULL)
+		luaL_error(l, "gk: failed to convert an IP address to string at %s!",
+			__func__);
+
+	ret = snprintf(fib_info, sizeof(fib_info),
+		"FIB entry for IP prefix: %s/%hhu with action %hhu\n",
+		ip, iface->ip4_addr_plen, neigh_fib->action);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(fib_info));
+
+	lua_pushstring(l, fib_info);
+	lua_concat(l, 2);
```

Comment:
This is another place that the proposed `struct gk_neighbor_dump_entry` will improve our code since we won't restrict Lua to operate on strings only. This comments also affects `list_ipv6_if_neighbors()`.

Reply:
The code above is dealing with the prefix information. The proposed ``struct gk_neighbor_dump_entry`` will deal with a single ethernet cache, it's possible that the FIB contains more than one cache entries. I wonder how would the proposed data struct improve the code here?

Reply:
It improves avoiding specifying that the type of the variable that the Lua callback works with is a string, that is, the accumulator can have any type.

I forgot to add the prefix information to `struct gk_neighbor_dump_entry`.

Reply:
I see your point. Since the lua function ``print_neighbor_dump_entry()`` will be called for each ethernet cache, and even 0 cache entry. I think we may have a way to print the prefix information at the first call of ``print_neighbor_dump_entry()``.

Reply:
Notice that passing the prefix during the first call of `print_neighbor_dump_entry()` isn't enough because there are front and back neighbors, that is, there are two prefixes. In the interface variations that I could come up with to isolate the prefixes, we end up complicating the callback interface. For example, adding parameters to `print_neighbor_dump_entry()` or adding a second Lua callback just to deal with the prefix.

The current code was motivated to support diagnosing of issues in production. And, in production, having no neighbors is not a viable setup. Even if a no neighbor condition is triggered, someone could still add a non-existing neighbor to produce a stale entry to test things up. So it seems to me that we don't have a reason to complicate the interface for now.

What do you think?

Reply:
It's fine to not include the prefix in `struct gk_neighbor_dump_entry`. Moreover, we don't need to add one more field to have the interface information since the field `action` already provides it.